From 1ba27de00d083a2c77ba4a329ab06b9ab748e9b8 Mon Sep 17 00:00:00 2001
From: Peter van der Perk <peter.vanderperk@nxp.com>
Date: Fri, 20 Jun 2025 15:14:38 +0200
Subject: [PATCH] qemu-Reinstated-syscall-emulation-through-libc-8

Upstream-Status: Inappropriate [embedded specific]

---
diff --git a/linux-user/main.c b/linux-user/main.c
index c49d1e9..f09c158 100644
--- a/linux-user/main.c
+++ b/linux-user/main.c
@@ -429,6 +429,11 @@ static void handle_arg_strace(const char *arg)
     enable_strace = true;
 }
 
+static void handle_arg_libcsyscall(const char *arg)
+{
+    do_libcsyscall = 1;
+}
+
 static void handle_arg_version(const char *arg)
 {
     printf("qemu-" TARGET_NAME " version " QEMU_FULL_VERSION
@@ -518,6 +523,8 @@ static const struct qemu_argument arg_table[] = {
      "size",       "TCG translation block cache size"},
     {"strace",     "QEMU_STRACE",      false, handle_arg_strace,
      "",           "log system calls"},
+    {"libcsyscall","QEMU_LIBCSYSCALL", false, handle_arg_libcsyscall,
+     "",           "run system calls for fakeroot/pseudo through libc"},
     {"seed",       "QEMU_RAND_SEED",   true,  handle_arg_seed,
      "",           "Seed for pseudo-random number generator"},
     {"trace",      "QEMU_TRACE",       true,  handle_arg_trace,
diff --git a/linux-user/syscall.c b/linux-user/syscall.c
index 7404af9..3daed0a 100644
--- a/linux-user/syscall.c
+++ b/linux-user/syscall.c
@@ -66,6 +66,7 @@
 #include <linux/in6.h>
 #include <linux/errqueue.h>
 #include <linux/random.h>
+#include <signal.h>
 #ifdef CONFIG_TIMERFD
 #include <sys/timerfd.h>
 #endif
@@ -163,6 +164,8 @@
 #define TCSETSF2 TCSETSF
 #endif
 
+int do_libcsyscall=0;
+
 #ifndef CLONE_IO
 #define CLONE_IO                0x80000000      /* Clone io context */
 #endif
@@ -283,6 +286,37 @@ static type name (type1 arg1,type2 arg2,type3 arg3,type4 arg4,type5 arg5,	\
 	return syscall(__NR_##name, arg1, arg2, arg3, arg4, arg5, arg6);	\
 }
 
+#define _libc_syscall1(type,name,namelibc,type1,arg1)  \
+static type name (type1 arg1)        \
+{              \
+	if (do_libcsyscall)       \
+		return namelibc(arg1);      \
+	return syscall(__NR_##name, arg1);    \
+}
+
+#define _libc_syscall2(type,name,namelibc,type1,arg1,type2,arg2)	\
+static type name (type1 arg1,type2 arg2)				\
+{									\
+ if (do_libcsyscall)							\
+   return namelibc(arg1, arg2);						\
+ return syscall(__NR_##name, arg1, arg2);				\
+}
+
+#define _libc_syscall3(type,name,namelibc,type1,arg1,type2,arg2,type3,arg3)  \
+static type name (type1 arg1,type2 arg2,type3 arg3)        \
+{                    \
+ if (do_libcsyscall)             \
+   return namelibc(arg1, arg2, arg3);        \
+ return syscall(__NR_##name, arg1, arg2, arg3);        \
+}
+
+#define _libc_syscall4(type,name,namelibc,type1,arg1,type2,arg2,type3,arg3,type4,arg4)	\
+static type name (type1 arg1,type2 arg2,type3 arg3,type4 arg4)				\
+{										\
+	if (do_libcsyscall)							\
+		return namelibc(arg1, arg2, arg3, arg4);				\
+	return syscall(__NR_##name, arg1, arg2, arg3, arg4);				\
+}
 
 #define __NR_sys_uname __NR_uname
 #define __NR_sys_getcwd1 __NR_getcwd
@@ -478,13 +512,12 @@ static const bitmask_transtbl fcntl_flags_tbl[] = {
 #endif
 };
 
-_syscall2(int, sys_getcwd1, char *, buf, size_t, size)
+_libc_syscall2(int, sys_getcwd1, getcwd, char *, buf, size_t, size)
 
 #if defined(TARGET_NR_utimensat) || defined(TARGET_NR_utimensat_time64)
 #if defined(__NR_utimensat)
 #define __NR_sys_utimensat __NR_utimensat
-_syscall4(int,sys_utimensat,int,dirfd,const char *,pathname,
-          const struct timespec *,tsp,int,flags)
+_libc_syscall4(int,sys_utimensat,utimensat,int,dirfd,const char*,pathname,const struct timespec *,times,int,flags);
 #else
 static int sys_utimensat(int dirfd, const char *pathname,
                          const struct timespec times[2], int flags)
@@ -649,6 +682,31 @@ int copy_struct_from_user(void *dst, size_t ksize, abi_ptr src, size_t usize)
     return 0;
 }
 
+#define libc_syscallargs(name, ...) \
+( \
+    (do_libcsyscall) ? \
+        name(__VA_ARGS__) \
+    : \
+        safe_syscall(__NR_##name, __VA_ARGS__) \
+)
+
+#define libc_syscall3(type, name, type1, arg1, type2, arg2, type3, arg3) \
+static type safe_##name(type1 arg1, type2 arg2, type3 arg3) \
+{ \
+    if (do_libcsyscall) \
+        return name(arg1, arg2, arg3); \
+    return safe_syscall(__NR_##name, arg1, arg2, arg3); \
+}
+
+#define libc_syscall4(type, name, type1, arg1, type2, arg2, type3, arg3, \
+    type4, arg4) \
+static type safe_##name(type1 arg1, type2 arg2, type3 arg3, type4 arg4) \
+{ \
+    if (do_libcsyscall) \
+        return name(arg1, arg2, arg3, arg4); \
+    return safe_syscall(__NR_##name, arg1, arg2, arg3, arg4); \
+}
+
 #define safe_syscall0(type, name) \
 static type safe_##name(void) \
 { \
@@ -698,7 +756,7 @@ static type safe_##name(type1 arg1, type2 arg2, type3 arg3, type4 arg4, \
 
 safe_syscall3(ssize_t, read, int, fd, void *, buff, size_t, count)
 safe_syscall3(ssize_t, write, int, fd, const void *, buff, size_t, count)
-safe_syscall4(int, openat, int, dirfd, const char *, pathname, \
+libc_syscall4(int, openat, int, dirfd, const char *, pathname, \
               int, flags, mode_t, mode)
 
 safe_syscall4(int, openat2, int, dirfd, const char *, pathname, \
@@ -710,7 +768,7 @@ safe_syscall4(pid_t, wait4, pid_t, pid, int *, status, int, options, \
 #endif
 safe_syscall5(int, waitid, idtype_t, idtype, id_t, id, siginfo_t *, infop, \
               int, options, struct rusage *, rusage)
-safe_syscall3(int, execve, const char *, filename, char **, argv, char **, envp)
+libc_syscall3(int, execve, const char *, filename, char **, argv, char **, envp)
 safe_syscall5(int, execveat, int, dirfd, const char *, filename,
               char **, argv, char **, envp, int, flags)
 #if defined(TARGET_NR_select) || defined(TARGET_NR__newselect) || \
@@ -818,9 +876,9 @@ safe_syscall4(int, fchmodat2, int, dfd, const char *, filename,
  * we should be using the 64-bit structures automatically.
  */
 #ifdef __NR_fcntl64
-#define safe_fcntl(...) safe_syscall(__NR_fcntl64, __VA_ARGS__)
+#define safe_fcntl(...) libc_syscallargs(fcntl64, __VA_ARGS__)
 #else
-#define safe_fcntl(...) safe_syscall(__NR_fcntl, __VA_ARGS__)
+#define safe_fcntl(...) libc_syscallargs(fcntl, __VA_ARGS__)
 #endif
 
 static inline int host_to_target_sock_type(int host_type)
@@ -7565,10 +7623,10 @@ static inline int tswapid(int id)
 #define __NR_sys_setregid __NR_setregid
 #endif
 
-_syscall1(int, sys_setuid, uid_t, uid)
-_syscall1(int, sys_setgid, gid_t, gid)
-_syscall3(int, sys_setresuid, uid_t, ruid, uid_t, euid, uid_t, suid)
-_syscall3(int, sys_setresgid, gid_t, rgid, gid_t, egid, gid_t, sgid)
+_libc_syscall1(int, sys_setuid, setuid, uid_t, uid)
+_libc_syscall1(int, sys_setgid, setgid, gid_t, gid)
+_libc_syscall3(int, sys_setresuid, setresuid, uid_t, ruid, uid_t, euid, uid_t, suid)
+_libc_syscall3(int, sys_setresgid, setresgid, gid_t, rgid, gid_t, egid, gid_t, sgid)
 _syscall2(int, sys_setgroups, int, size, gid_t *, grouplist)
 _syscall2(int, sys_setreuid, uid_t, ruid, uid_t, euid);
 _syscall2(int, sys_setregid, gid_t, rgid, gid_t, egid);
@@ -13371,9 +13429,18 @@ static abi_long do_syscall1(CPUArchState *cpu_env, int num, abi_long arg1,
                 }
                 tsp = ts;
             }
-            if (!arg2)
-                ret = get_errno(sys_utimensat(arg1, NULL, tsp, arg4));
-            else {
+            if (!arg2) {
+                /* Special solution to overcome a problem for the libc
+                 * based implementation, so that we don't get EINVAL.
+                 * libc is a bit restrictive there and non-Linux'ish.
+                 * The need is a Linux'ism as it seems.
+                 */
+               if (do_libcsyscall)
+                    ret = get_errno(futimens(arg1, tsp));
+                else
+                    ret = get_errno(sys_utimensat(arg1, NULL, tsp, arg4));
+	    }
+	    else {
                 if (!(p = lock_user_string(arg2))) {
                     return -TARGET_EFAULT;
                 }
diff --git a/linux-user/user-internals.h b/linux-user/user-internals.h
index 24d3599..6d20949 100644
--- a/linux-user/user-internals.h
+++ b/linux-user/user-internals.h
@@ -98,6 +98,7 @@ void probe_guest_base(const char *image_name,
 
 /* syscall.c */
 int host_to_target_waitstatus(int status);
+extern int do_libcsyscall;
 
 #ifdef TARGET_I386
 /* vm86.c */
-- 
2.34.1
